crates/libcgroups/src/systemd/dbus_native/dbus.rs

```diff
@@ -296,20 +303,47 @@ impl DbusConnection {
             None,
         )?;
 
-        let reply = self.receive_complete_response()?;
-
-        // note that a single received response can contain multiple
-        // messages, so we must deserialize it piece by piece
         let mut ret = Vec::new();
-        let mut buf = &reply[..];
-
-        while !buf.is_empty() {
-            let mut ctr = 0;
-            let msg = Message::deserialize(&buf[ctr..], &mut ctr)?;
-            // we reset the buf, because I couldn't figure out how the adjust_counter function
-            // should should be changed to work correctly with non-zero start counter, and this solved that issue
-            buf = &buf[ctr..];
-            ret.push(msg);
+
+        // it is possible that while receiving messages, we get some extra/previous message
+        // for method calls, we need to have an error or method return type message, so
+        // we keep looping until we get either of these. see https://github.com/containers/youki/issues/2826
+        // for more detailed analysis.
+        loop {
+            let reply = self.receive_complete_response()?;
+
+            // note that a single received response can contain multiple
+            // messages, so we must deserialize it piece by piece
+            let mut buf = &reply[..];
+
+            while !buf.is_empty() {
+                let mut ctr = 0;
+                let msg = Message::deserialize(&buf[ctr..], &mut ctr)?;
+                // we reset the buf, because I couldn't figure out how the adjust_counter function
+                // should should be changed to work correctly with non-zero start counter, and this solved that issue
+                buf = &buf[ctr..];
+                ret.push(msg);
+            }
+
+            // in Youki, we only ever do method call apart from initial auth
+            // in case it is, we don't really have a specific message to look
+            // out of, so we take the buffer and break
+            if mtype != MessageType::MethodCall {
+                break;
+            }
+
+            // check if any of the received message is method return or error type
+            let return_message_count = ret
+                .iter()
+                .filter(|m| {
+                    m.preamble.mtype == MessageType::MethodReturn
+                        || m.preamble.mtype == MessageType::Error
+                })
+                .count();
+
+            if return_message_count > 0 {
+                break;
+            }
```

Comment:
I'd like suggestion for other ways to implement this. Ideally I feel that `send_message` function should not have to deal with the specific message type, but as we collect the response here, we need to account for it. Other way I can think is if we separate the send message and get response functions, so that the higher level calling functions can choose to loop over the get response.  But that means we would have to do two function calls at each place we are doing a send message call. Would that be a better way?

Reply:
> But that means we would have to do two function calls at each place we are doing a send message call. Would that be a better way?

I do not think it is. The complexity of should be contained in `send_message.` The caller should not have to think about it.